# Data Analytics — Exercise 7.1: NumPy Essentials
## Cris Ramirez | Date: 05/19/2026
This notebook uses NumPy to analyse weekly revenue data for five regional offices.
It covers array creation, descriptive statistics, arithmetic operations, Boolean masking, and 2-D reshaping.

## 1. Import NumPy and Create the Revenue Array
We import NumPy and create a 1-D array of weekly revenue figures.

In [1]:
import numpy as np

revenue = np.array([52_400.0, 38_750.0, 61_200.0, 47_900.0, 55_650.0])
print(f'Array : {revenue}')
print(f'dtype : {revenue.dtype}')
print(f'shape : {revenue.shape}')

Array : [52400. 38750. 61200. 47900. 55650.]
dtype : float64
shape : (5,)


## 2. Descriptive Statistics
NumPy built-in functions compute summary statistics in one line each.

In [2]:
print('── Revenue Statistics ──────────────────────')
print(f'Total   : ${np.sum(revenue):>12,.2f}')
print(f'Mean    : ${np.mean(revenue):>12,.2f}')
print(f'Median  : ${np.median(revenue):>12,.2f}')
print(f'Std Dev : ${np.std(revenue):>12,.2f}')
print(f'Minimum : ${np.min(revenue):>12,.2f}')
print(f'Maximum : ${np.max(revenue):>12,.2f}')

── Revenue Statistics ──────────────────────
Total   : $  255,900.00
Mean    : $   51,180.00
Median  : $   52,400.00
Std Dev : $    7,578.03
Minimum : $   38,750.00
Maximum : $   61,200.00


## 3. Element-Wise Arithmetic — 5% Bonus
Multiplying a NumPy array by a scalar applies the operation to every element.

In [4]:
revenue_bonus = revenue * 1.05
bonus_amount = revenue_bonus - revenue

print('Original Revenue :', revenue)
print('Revenue + 5% Bonus:', revenue_bonus)
print('Bonus Amount      :', bonus_amount)

Original Revenue : [52400. 38750. 61200. 47900. 55650.]
Revenue + 5% Bonus: [55020.  40687.5 64260.  50295.  58432.5]
Bonus Amount      : [2620.  1937.5 3060.  2395.  2782.5]


## 4. Boolean Masking — Above-Average Regions
A Boolean mask filters the array to show only values meeting a condition.

In [6]:
avg = np.mean(revenue)
above_avg = revenue > avg

print(f'Mean       : ${avg:,.2f}')
print('Mask (T/F) :', above_avg)
print('Above average :', revenue[above_avg])
print(f'Regions above : {np.sum(above_avg)} of {len(revenue)}')

Mean       : $51,180.00
Mask (T/F) : [ True False  True False  True]
Above average : [52400. 61200. 55650.]
Regions above : 3 of 5


## 5. Reshape and 2-D Analysis
Reshaping allows us to view the data as a matrix and compute axis-based stats.

In [7]:
col_vec = revenue.reshape(5, 1)
print('Column vector (5,1):\n', col_vec)

q_data = np.vstack([revenue, revenue * 1.05])
print('\n2-D array (2 weeks × 5 regions):\n', q_data)

print(f'\nWeekly totals    : {np.sum(q_data, axis=1)}')
print(f'Region avg (2wk) : {np.mean(q_data, axis=0)}')

Column vector (5,1):
 [[52400.]
 [38750.]
 [61200.]
 [47900.]
 [55650.]]

2-D array (2 weeks × 5 regions):
 [[52400.  38750.  61200.  47900.  55650. ]
 [55020.  40687.5 64260.  50295.  58432.5]]

Weekly totals    : [255900. 268695.]
Region avg (2wk) : [53710.   39718.75 62730.   49097.5  57041.25]


## Summary
- The five regions generated a combined total of 256,900 in Week 1.
- Region 3 (61,200) is the top performer; Region 2 (38,750) is the lowest.
- 3 of 5 regions exceeded the mean of 51,380.
- Applying the 5% bonus raises the Week 2 total to 269,695.